cmdline: new module for building CLIs

alaviss opened the pull request:
## Summary

Introducing  `cmdline` , a new experimental module aiming to simplify
the process of building command line interfaces.

## Details

The standard library currently lacks an option for quickly spinning up
command line tools. While  `parseopt`  and  `lexopt`  can handle the
process of reading command line parameters, they are not equipped to:

- Generate help messages for a command
- Parse the command line corresponding to a declared interface
- Automatically handle errors in user input and print helpful
diagnostic

`cmdline`  is designed to fulfill those requirements, and to become the
standard method to construct command line interfaces.

A key part in the design of this module is to facilitate support for
other frontends. Many applications now feature an interface that
provides:

- Colored errors
- Colored help messages
- Producing auto correct suggestions on errors
- etc.

While  `cmdline`  can implement all of these features on its own, it
would balloon the complexity of the module beyond what is reasonable for
a standard library component (factoring in the current maintainer
bandwidth). As such a few design constraints was applied:

- The core structures ( `Cli`  and  `parse()` ) are tightly coupled for
both simplicity and to enforce a standardized interpretation.
- Core structures avoid the use of meta programming: a  `Cli`  must be
able to be built using only  `proc` s. This provide a stable basis for
sugar like DSLs and automatic generators to work with.
- Auxiliary functions like  `help()` ,  `run()`  are implemented using
**only** public interfaces, making sure that user can bring their own
implementation. Additionally, the components that generate portions of
help messages/diagnostics are also made public, allowing a custom
composite to be made.

alaviss committed:
cmdline: new module for building CLIs

alaviss committed:
cmdline: implement aliases

alaviss committed:
cmdline: tests for docgen, aliases and some code cleanup

alaviss committed:
cmdline: implement positional parameters

alaviss committed:
cmdline: rewrite parser as tailcalls

This is a preparation for adding command support.

alaviss committed:
cmdline: switch to stdlib utilities and general cleanup

alaviss committed:
cmdline: implement subcommand support

alaviss committed:
cmdline: don't expose positional position

Require defining how position is calculated without a lot of benefits

alaviss committed:
cmdline: add support for callbacks with no action

alaviss committed:
cmdline: implement runner

saem commented on lib/experimental/cmdline/parsers.nim:
```diff
@@ -0,0 +1,52 @@
+import std/strutils
+import std/typetraits
+
+proc parseCli*(T: typedesc[SomeInteger], value: string): T =
```
`tryParseCli` variants that don't raise but instead return a default value (taken as a parameter) would be handy, as people will be doing a lot of that.

alaviss replied:
I'm not sure about the value of that for CLI purposes. More often than not you want to error out because wrong inputs from user meant that they make a mistake and shouldn't be ignored. For general parsing `parseutils` is what you want to use.

Other than that, a `try` variant never make much sense to me, as it'd require you to always construct a (potentially expensive) object regardless of parsing results. I could entertain the idea of using `Result[T]`, which would come with free support for defaults, but the ecosystem at the moment deals in exceptions.

saem replied:
The idea with try variants is that you use in band error values to avoid the crappiness of `try...except` wrapping.

`Result[T]` is better, I keep forgetting we have one in the stdlib/experimental (I remember it as for the compiler). 😆

alaviss replied:
I elected to not use `Result[T]` for now.

Most of the APIs an user would call to aid with parsing raises `ValueError` themselves, and at present this makes it a lot simpler to write a custom `parseCli` implementation, which I expect to happen more than explicitly invoking these.

We could add a capturing feature to `Result[T]` which would automatically turn exceptions into an error `Result`, which then an user can just do something like:

```nim
let val = capture(parseCli(int, "not int")).valueOr(-1)
```

and would work for all existing APIs, not just this one.

saem commented:
Not sure what your plan for docs is, but I recommend only covering the public elements and making it largely runnable examples. I've found the lack of the latter has been the biggest gap in docs. Not to mention it helps with setting us up for LLMs, because they're so example driven.

alaviss committed:
cmdline: basic description of APIs

alaviss committed:
cmdline: make sure that changing name removes duplicates from alias

alaviss committed:
cmdline: more docs and a quick feature addition

added a defaults param for accumulator-less run/parse

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,1945 @@
+    # Might be useful to support not having these for space-constrained
+    # targets using a define.
+    usage: Store[Parameter, string] ## Canonical usage for Parameters.
+    placeholder: Store[Parameter, string] ## Canonical placeholder for
+                                          ## Parameters. Only used for flags.
```
Not a blocker, but why not put the doc comment on the next line? Re-formating the text in case of name/type changes becomes simpler that way.

alaviss replied:
Force of habit

saem commented on doc/lib.rst:
```diff
@@ -386,6 +383,18 @@ Parsers
   to be somewhat error-correcting, so that even some "wild HTML" found on the
   web can be parsed with it.
 
+Command line parsers
```
```suggestion
Command Line Parsers
```

Please use title case for titles

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2205 @@
+
+type
+  Cli*[T] {.requiresInit.} = object
+    ## A command line interface.
```
```suggestion
    ## A command line interface description.
```

It's a bit more accurate

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2205 @@
+    Flag
+    Positional
+
+  Parameter* = distinct uint32
```
```suggestion
  ParameterId* = distinct uint32
```

I would suffix these with `Id`, so it's clear that they need to be passed to something in order to be useful/get more information, etc.

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2205 @@
+  result.aliases.setLen(0)
+
+  # Not the fastest method, but it's expected that users will
+  # specify at most 4 of these.
```
```suggestion
  # Not the fastest method, but it's assumed that users will
  # specify a handful (approximately 1-4) of these.
```

`assumed`, unlike `expected`, doesn't imply we might, or could in the future, enforce this

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2205 @@
+    command: Table[Command, CliCommand] ## Lookup mapping of command to lookup tables.
+    parser: Store[Parameter, ParserAny[T]] ## Parser to process input for Parameter.
+    name: Store[Parameter, string] ## Canonical names for all Parameters.
+    alias: Table[Parameter, seq[string]] ## Mapping of Parameter to aliases.
```
Just so it's more self documenting and avoids some potential gotchas, we should probably use a set like structure instead of a `seq` to store the list of aliases, to avoid duplicates.

alaviss replied:
We don't have any such thing at the moment :disappointed: 

Right now the places that modify the table actually perform the required checks, so I'd say that it should be fine.

saem replied:
Bummer, we could really use a set like that. Although, sets push the need for interfaces pretty hard, because there are so many different implementations. 😬

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2805 @@
+  messageOutput: File = stdmsg,
+) =
+  ## Parses the command line, collecting changes made by added parsers
+  ## to `accumulator`.
```
```suggestion
  ## Parses the command line `args` based on the description in `cli`, with configured
  ## parsers updating values in the `accumulator`.
```

Makes it a little bit more clear as to what's going on.

saem reviewed (commented):
Did another pass and came up with some suggestions

alaviss committed:
cmdline: apply review suggestions

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+# See the file "copying.txt", included in this distribution, for
+# details about copyright.
+
+from os import commandLineParams
```
Could you remove the usage of `commandLineParams`?

It would allow using `cmdline` on a broader set of targets, such as NimScript, standalone VM execution (`nim vm`), and JavaScript code running in the browser. Ideally, no such concessions should be necessary to support other backends, but at the moment, they are.

Separate from the above, I also think not a good idea to hide where the command line comes from (by using `commandLineParams()` as the default value for `run.args`).

alaviss replied:
I think having the default makes usage a lot simpler. `run()` is meant to take the wheel and handle all the boring stuff for you, including process termination, and is relatively simple on it's own. For a more controlled experience, `parse()` is what you want.

I could use a `when` to remove `commandLineParams` as default for targets without command line parameters, or simply remove `run` for them altogether, if that works.

saem replied:
Dropping the dependency upon `commandLIneParams` is a must imo, and `when`, even though its a compromise, at least puts us in the right direction.

This also let's us make sure we can include this in stress testing the js backend.

zerbina replied:
A perhaps cleaner solution compared to using `when` is moving `run` into its own module, where it can freely depend upon `os`, with a possible candidate for the name being `cmdlinerunner`.

In my opinion, `run` is part of the "porcelain layer", and thus doesn't really belong with the rest of `cmdline`. Having it separate also makes sure that the implementation doesn't start accidentally rely on private `cmdline` APIs.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+import cmdline/parsers
+export parsers
+
+import lexopt
```
```suggestion
import
  std/[
    algorithm,
    hashes,
    options,
    sequtils,
    strutils,
    tables
  ],
  std/private/[
    containers
  ],
  cmdline/[
    parsers
  ],
  experimental/[
    lexopt
  ]

export parser
```
It's a tad longer, but new code generally uses import sections and blocks.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+import lexopt
+
+## This module implements a fast and lightweight declarative command line
```
Please move the module-level documentation above the imports.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## See also:
+    ## - `initCli proc <#initCli,sinkCommandBuilder[T]>`_
+    command: Table[CommandId, CliCommand]
+      ## Lookup mapping of command to lookup tables.
```
These comments feel a bit redundant -- the type already shows what the field is meant for. I also think the fields here should use plural (e.g., `commands` instead of `command`).

alaviss replied:
> These comments feel a bit redundant -- the type already shows what the field is meant for.

The comments are mostly seeds for additional details (if needed) in the future. I could drop them if wanted, of course.

> I also think the fields here should use plural (e.g., `commands` instead of `command`).

I think they are better in singular form, the reason being while they technically store multiples of X, the usage of them is usually a direct binding with an ID, so I prefer to name them as if I'm dealing with a single object (and if we ever have a SoA feature this comes in handy).

saem replied:
> > I also think the fields here should use plural (e.g., `commands` instead of `command`).
> 
> I think they are better in singular form, the reason being while they technically store multiples of X, the usage of them is usually a direct binding with an ID, so I prefer to name them as if I'm dealing with a single object (and if we ever have a SoA feature this comes in handy).

This isn't in keeping in the style with how we've been writing things and I'm not seeing a good reason to deviate, tbh. If it's many, it should be plural. Reading through the code it is weird that we have to continually provide a `CommandId` to `cli.command`.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+  FlagParser*[T; R: MaybeAction] = proc (option, value: string, accumulator: var T): R
+    ## A parser for flag with name `option` and value `value`. The accumulator
+    ## passed to `run()`_ or `parse()`_ can be accessed and modified via
```
```suggestion
    ## passed to `run`_ or `parse`_ can be accessed and modified via
```
Could you remove the parenthesis? I don't think provide any benefit to clarity, especially given that the link is a local one.

(The same goes for other run()_ and parse()_ occurrences below)

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## passed to `run()`_ or `parse()`_ can be accessed and modified via
+    ## `accumulator`.
+    ##
+    ## A parser may choose to return an `Action` or not return anything,
```
I known that inlining is not possible, but the doc comment duplication is a bit much, and I'd prefer if the expectations of a parser are put under a "Writing Parsers" (or something along that line) section in the top-level documentation, to which every parser type then links.

Autocomplete user won't have immediate access to the information then (unless using an IDE that supports following links in documentation boxes), but I don't think it's super important for the types.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## processed accordingly by the library. Any other exceptions signify
+    ## an internal error and will not be handled automatically.
+    ##
+    ## .. _run(): #run,Cli[T],T,sinkseq[string],File
+    ## .. _parse(): #parse,Cli[T],T,sinkseq[string]
```
```suggestion
    ## .. _run: #run,Cli[T],T,sinkseq[string],File
    ## .. _parse: #parse,Cli[T],T,sinkseq[string]
```

zerbina commented on doc/lib.rst:
```diff
@@ -386,6 +383,18 @@ Parsers
+  providing an easy to use solution for command line interfaces.
+
+* `experimental/lexopt <lexopt.html>`_
+  The `lexopt` module implements an user-driven command-line option lexer.
```
```suggestion
  The `lexopt` module implements a user-driven command-line option lexer.
```

zerbina reviewed (commented):
I've gathered some first suggestions/remarks, mostly regarding style, some typos, and documentation, but I haven't looked at all code yet.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## Lookup table for flags and positionals.
+    flag: Table[string, ParameterId]
+      ## Lookup mapping of flag names to ParameterId.
+    command: Table[string, CommandId]
```
I think it'd be clearer if the field is called `subcommands`. At least the documentation should mention that it stores the subcommands.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ##
+    ## See also:
+    ## - `positionalBuilder proc <#positionalBuilder,Cli[T]>`_
+    posParser: ParserAny[T]
```
Could you use `PositionalParser` directly, with boolean fields for catch-all and optional? The implementation would become clearer and simpler, in my opinion.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ##
+    ## See also:
+    ## - `commandBuilder proc <#commandBuilder,Cli[T]>`_
+    cmdParser: ParserAny[T]
```
The comment above also applies here, just with `CommandParser`.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## Action to be taken after parsing.
+    Continue ## Continue parameter parsing.
+    ShowHelp ## Abort and show help message.
+    DisableFlagProcessing ## ParameterIds following this will be considered
+                          ## to be values.
```
The comment is in conflict with the name. The name says that only flag processing is disabled, whereas the documentation says that commands, positionals, and flags will be considered values (presumably for flags).

alaviss replied:
Yea that's outdated docs back when only flags are supported. And apparently it also got caught by my substitution filter for the `Id` suffix the other day.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## Typeclass to support parsers returning either `Action` or nothing.
+
+  FlagParser*[T; R: MaybeAction] = proc (option, value: string, accumulator: var T): R
+    ## A parser for flag with name `option` and value `value`. The accumulator
```
I don't get the point of having the `option` parameter. Wouldn't a separate procedure be used instead of dispatching on the `option` within the procedure?

From what I've seen, none of the tests and runnable examples actually make use of `option`, too.

alaviss replied:
> Wouldn't a separate procedure be used instead of dispatching on the `option` within the procedure?

It allows for implementing things like `--no-foo` vs `--foo` by aliasing on `--foo`.

It's generally useful to expose input values whenever possible IMO, there's always an usage for someone and it's not hidden internal information.

zerbina replied:
The information isn't really hidden, as one explicitly provides the names/aliases the parsers acts on, but I do get your point.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## .. _parse(): #parse,Cli[T],T,sinkseq[string]
+
+  CommandParser*[T; R: MaybeAction] = proc (command: CommandId, accumulator: var T): R
+    ## A parser for command parameter with command `command`. The accumulator
```
From what I can tell, a `CommandParser` doesn't actually parse anything in the common sense of the word. Rather, it's meant to *process* a command appearing.

Personally, I think using a different name would be better. "processor" could work, although "command processing" usually has a slightly different meaning.

alaviss replied:
The accumulator also acts as an input that could inform whether a command is legal at a given position. At least that's the excuse I would give :laughing:

The real reason is that I centered a lot of the vocab used around `parser()`, which makes for a consistent-ish API.

zerbina replied:
I'm not sure whether the regularity is something desired here, as the operation is conceptually (a little) different from the others, but the name is something that can still be easily adjusted later (if at all) -- the module is experimental, after all.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    ## .. _run(): #run,Cli[T],T,sinkseq[string],File
+    ## .. _parse(): #parse,Cli[T],T,sinkseq[string]
+
+  FlagTypedParser*[T; U; R: MaybeAction] = proc (option: string, value: U, accumulator: var T): R
```
Using "processor" for the name would fit better, I think, as the actual parsing is delegated to `parseCli`.

alaviss replied:
While `parseCli` would be doing parsing here, it's not unlikely that the user might perform additional verification based on the collected state so far, which is a form of parsing to me :grin:

saem replied:
Reading through it some more, I think `extractor` is a better term, in that it's more precise, in describing what's going on based on the API + examples/tests I've read thus far, it's extracting a value.

N.B.: for the builder api, `extract` reads better, e.g.: `.extract(noop)`.

alaviss replied:
I think for API consistency I'll stick with `Parser` here. It's easier to convey these as "parser" and "typed parser", as they both put data into the accumulator. As a bonus, APIs that can take both will clearly signal that they are exclusive and will override the other (i.e. `flagBuilder().parser().parser(bool)` vs. `flagBuilder().parser().extractor()`).

saem replied:
I personally don't see putting data into an accumulator as a key characteristic of a parser, the first, and predominant, thing that comes to mind is screwing around with character/token streams.

alaviss replied:
The point is that I'd prefer to use only one term to refer to all of these. I think `parser` fits this role best, but I could be convinced otherwise.

saem replied:
I'm ok to give it a shot and see what it feels like when folks, including myself, use it and then come back to this, it's not like we guarantee stability yet, doubly so for anything experimental.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    CatchAll
+    OptionalCatchAll
+
+  ParserAny[T] = object
```
Have you considered only using `FlatOptionalParser` internally? `FlagBuilder.parser` would create a wrapper, meaning there would be no change in public interface.

The benefit is that you then don't need to use `ParserAny` in `FlagBuilder`, which would improve clarity of the code by a good bit, in my opinion.

alaviss replied:
It's not possible since we need reflection information for `isValueOptional`. And also the parsing routine is ill-equipped to handle `ParseError` generated by user-provided parsers.

The other motivation was to avoid closures, but that ship sailed the moment I allow parsers to not return `Action`.

zerbina replied:
> It's not possible since we need reflection information for `isValueOptional`.

Hm, why? `FlagOptionalValue` wouldn't go away -- there would just be no `FlagParser[T, Action]` field in `ParserAny`.

alaviss replied:
Oh, right, my bad. But the other reason stands, since error handling for no value case would have to be done outside of the actual parser, as the facilities for handling parser-raised `ParseError` doesn't exist and it would force exposing certain internal states.

I could invent a new "internal" exception for this purpose, but that wouldn't change anything other than where the handling code is, and the current method prevents unnecessary code duplication.

An alternative to just unwrap unconditionally after outside checks, but I'd prefer a different prototype to avoid internal bugs.

zerbina replied:
I'm not entirely following, sorry. As far as I can tell, `optParser` and `parser` are only called in `parseFlag`, and therefore having only `optParser` would only require changing changing the `parser.parser(option, ...)` call to take the option directly (plus creating a wrapper in `FlagBuilder.parser`).

In any case, looking at the code again, I don't think this would result in a noticeable improvement.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    of ParserKind.Flag: parser: FlagParser[T, Action]
+    of ParserKind.Positional, OptionalPositional, CatchAll,
+       OptionalCatchAll: posParser: PositionalParser[T, Action]
+    of ParserKind.Command: cmdParser: CommandParser[T, Action]
```
```suggestion
    of FlagOptionalValue:
      optParser: FlagOptionalParser[T, Action]
    of ParserKind.Flag:
      parser: FlagParser[T, Action]
    of ParserKind.Positional, OptionalPositional, CatchAll,
       OptionalCatchAll:
      posParser: PositionalParser[T, Action]
    of ParserKind.Command:
      cmdParser: CommandParser[T, Action]
```
Longer, but arguably also easier to read.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## this flag on the command line.
+  ##
+  ## If `name` is one-character long, it can also be recognized on the command
+  ## line using the short form syntax (e.g. `-n`).
```
```suggestion
  ## If `name` is only one ASCII character long, it may use the short form
  ## syntax (e.g. `-n`).
```
A soft suggestion. It's a bit more concise and also makes it clear that non-ASCII Unicode character cannot use the short form.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## Sets aliases for this flag.
+  ##
+  ## The command can be matched using any of the provided `names` in addition to
+  ## its canonical name.
```
```suggestion
  ## The flag can be matched using any of the provided `names` in addition to
  ## its canonical name.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+func default*[T](b: sink CommandBuilder[T]): CommandBuilder[T] =
+  ## Marks this command as the default subcommand of the current dispatcher.
+  ## When no subcommand are specified on the command line, this command
```
```suggestion
  ## When no subcommand is present on the command line, this command
```
A soft suggestion. I think it's slightly easier to understand this way.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## will be selected.
+  ##
+  ## Only one default subcommand is permitted per command. Attempting to
+  ## add more than one is considered an error.
```
```suggestion
  ## Only one default subcommand is permitted per command. Attempting to
  ## add another default subcommand to a command will cause an error.
```
A soft suggestion. The new wording is more definitive.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  result.isDefault = true
+
+func catchAll*[T](b: sink PositionalBuilder[T]): PositionalBuilder[T] =
+  ## Marks this positional as "catch all". All positional parameters encountered
+  ## starting at this positional will be handled by the associated parser.
```
```suggestion
  ## Marks this positional as "catch all". All positional parameters
  ## appearing beyond this positional will be handled by the associated
  ## parser.
```
Soft suggestion.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  let parser =
+    case result.posParser.kind
+    of ParserKind.Command..FlagOptionalValue:
+      nil
```
If I'm reading the code correctly, this case is can never happen, no? If you don't opt for removing the `ParserAny` usage for `PositionalBuilder`, I think the case expression here should be removed.

alaviss replied:
The default value of `parser` would use `ParserKind.Command`. Handling them here removes the ordering constraint between modifiers and parsers.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  let parser =
+    case result.posParser.kind
+    of ParserKind.Command..FlagOptionalValue:
+      nil
```
The above comment also applies here.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  b: sink FlagBuilder[T],
+  p: sink FlagOptionalParser[T, Action],
+): FlagBuilder[T] =
+  ## Sets the parser for this flag, and mark the flag as not requiring any value.
```
```suggestion
  ## Sets the parser for this flag, and marks the flag as not requiring any
  ## value.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## Sets the parser for this flag, and mark the flag as not requiring any value.
+  ##
+  ## Flags with optional value will only receive their value when it is
+  ## specified inline, for example: `--flag=value` or `--flag:value`.
```
```suggestion
  ## For flags with an optional value, their parser will only be called when
  ## the value is specified inline, for example: `--flag=value` or
  ## `--flag:value`.
```
Soft suggestion. I think the meaning of "receive" is not entirely clear in this context.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  p: sink FlagOptionalParser[T, void],
+): FlagBuilder[T] =
+  ## Sets the parser for this flag, and mark the flag as not requiring any value.
+  ## On successful parse, `Action.Continue` is taken as the default action.
```
```suggestion
  ## Sets the parser for this flag with `Action.Command` as the default value,
  ## and marks the flag as not requiring any value.
```
For symmetry with `CommandBuilder.parser`.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## Sets the parser for this flag, and mark the flag as not requiring any value.
+  ## On successful parse, `Action.Continue` is taken as the default action.
+  ##
+  ## Flags with optional value will only receive their value when it is
+  ## specified inline, for example: `--flag=value` or `--flag:value`.
```
```suggestion
  ## For flags with an optional value, their parser will only be called when
  ## the value is specified inline, for example: `--flag=value` or
  ## `--flag:value`.
```
Soft suggestion. I think the meaning of "receive" is not entirely clear in this context.

alaviss replied:
This phrasing wouldn't work well, since the parsers are also called without a value (i.e. `--opt-int` calls `opt-int`'s parser with a `none(int)`).

I think "receive" works here, given that there are supplementary examples below.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+func parser*[T](b: sink FlagBuilder[T], p: sink FlagParser[T, void]): FlagBuilder[T] =
+  ## Sets the parser for this flag, and mark the flag as requiring values.
+  ## On successful parse, `Action.Continue` is taken as the default action.
```
```suggestion
  ## Sets the parser for this flag with `Action.Continue` as the default
  ## action, and marks the flag as requiring a value.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  _: typedesc[U],
+  parser: sink FlagOptionalTypedParser[T, U, R],
+): FlagBuilder[T] =
+  ## Sets the parser for this flag, and mark the flag as not requiring any value.
```
```suggestion
  ## Sets the parser for this flag, and marks the flag as not requiring any
  ## value.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## `parseCli` before handing off to the parser. See `parsers module`_ for
+  ## more information.
+  ##
+  ## Flags with optional value will only receive their value when it is
+  ## specified inline, for example: `--flag=value` or `--flag:value`.
```
```suggestion
  ## For flags with an optional value, the parser will only be called when
  ## the value is specified inline, for example: `--flag=value` or
  ## `--flag:value`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  _: typedesc[U],
+  parser: sink FlagTypedParser[T, U, R],
+): FlagBuilder[T] =
+  ## Set the parser for this flag, and mark the flag as requiring values.
```
```suggestion
  ## Sets the parser for this flag, and marks the flag as requiring values.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## require a value to be passed. If values are to be passed, it must be
+  ## inlined (i.e. `--flag=false`).
+  ##
+  ## Flags with required value can receive any of the following forms:
```
```suggestion
  ## Flags with a required value can receive any of the following forms:
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  b: sink PositionalBuilder[T],
+  usage: sink string,
+): PositionalBuilder[T] =
+  ## Sets a short `usage` description and a value `placeholder` for this flag.
```
```suggestion
  ## Sets a short `usage` description for this positional.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  discard cli.parser.add(parser)
+  discard cli.parent.add(parent)
+
+func initCli*[T](b: sink CommandBuilder[T]): Cli[T] =
```
```suggestion
func initCli*[T: not void](b: sink CommandBuilder[T]): Cli[T] =
```
You'll get a rather cryptic error when passing a `CommandBuilder[void]` instance to `initCli`, otherwise.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## - Have no `parser` set.
+  ## - Have no `aliases`.
+  ##
+  ## Unlike subcommands, the root command may:
+  ##
+  ## - Have no `name` set. The `name` of a root command is used for
+  ##   documentation generation and does not participate in parsing.
```
I don't understand this comment. What happens if a name *is* set?

alaviss replied:
I suppose I should rephrase this, I'd welcome suggestions if you have any.

What is meant here is that the root command doesn't have to be named (since the name is used for docgen only), while subcommands must have a name.

EDIT: to answer the question directly: A name on the root command doesn't participate in parsing, so nothing would happen if one is set, hence its optionality.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## - Have no `name` set. The `name` of a root command is used for
+  ##   documentation generation and does not participate in parsing.
+  ##
+  ## The root command can be referred to using `RootCommand` constant.
```
```suggestion
  ## The root command can be referred to using the `RootCommand` constant.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## - `command` does not contain any positional parameters.
+  ## - A non-empty `name` must be set.
+  ## - None of the names specified by either `name` or `alias` are added to
+  ##   `command` previously.
```
```suggestion
  ## - `command` has no positional parameters.
  ## - The subcommand has a non-empty name.
  ## - Neither the name nor aliases are used by previously added subcommands.
```
So that the items all read as predicates.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ##
+  ## See also:
+  ## - `commandBuilder proc <#commandBuilder,Cli[T]>`_
+  assert b.cmdName.len > 0, "Command name must not be empty"
```
According to the doc comment this should be an error. I think it's fine if an assertion is used, but then the documentation needs to be adjusted.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## - A non-empty `name` must be set.
+  ## - None of the names specified by either `name` or `alias` are added to
+  ##   `command` previously.
+  ## - A `parser` or `optionalParser` must be set.
```
```suggestion
  ## - The flag has a non-empty name.
  ## - Neither the name nor aliases are used by previously added flags.
  ## - A parser or optional parser was set for the flag.
```
So that the items all read as predicates.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ##
+  ## See also:
+  ## - `flagBuilder proc <#flagBuilder,Cli[T]>`_
+  assert b.flagName.len > 0, "Flag name must not be empty"
```
Should raise a `ValueError` according to the documentation.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    if alias in cli.command[command].flag:
+      raise newException(ValueError, "Flag '" & alias & "' already exists")
+
+  assert not b.flagParser.isNil(), "Parser must be non-nil"
```
Same here.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ##   previously.
+  ## - A `parser` must be set.
+  ## - If non-optional, must be added prior to any optional parameters.
+  ## - If catch all, must be added after all other parameters.
```
```suggestion
  ## - `command` does not contain any subcommands.
  ## - The parameter has a non-empty name.
  ## - The name is not used by any previously added positional parameters.
  ## - A parser was set for the parameter.
  ## - (only when non-optional) No optional parameters have been added yet.
  ## - (only when not catch-all) No catch-all parameter have been added yet.
```
So that all items read as predicates.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ##
+  ## See also:
+  ## - `positionalBuilder proc <#positionalBuilder,Cli[T]>`_
+  assert b.posName.len > 0, "Positional name should not be empty"
```
Should raise a `ValueError` according to the documentation.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+      raise newException(ValueError):
+        "Positional with name '" & cli.name[param] & "' already exists"
+
+  assert not b.posParser.isNil(), "Parser must be non-nil"
```
Should raise a `ValueError` according to the documentation.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+iterator namesOf*(cli: Cli, flag: FlagId): lent string =
+  ## Yields names that can be used to refer to `flag`.
+  ##
+  ## The first name yielded is always the canonical name of `flag`.
```
```suggestion
  ## Returns all names that can be used to refer to `flag`. The canonical name
  ## is always returned first.
```
A soft suggestion.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+iterator namesOf*(cli: Cli, command: CommandId): lent string =
+  ## Yields names that can be used to refer to `command`.
+  ##
+  ## The first name yielded is always the canonical name of `command`.
```
```suggestion
  ## Returns all names that can be used to refer to `command`. The canonical
  ## name is always returned first.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  case cli.parser[param].kind
+  of ParserKind.Command: ParameterKind.Command
+  of ParserKind.Flag, FlagOptionalValue: ParameterKind.Flag
+  of ParserKind.Positional..OptionalCatchAll: ParameterKind.Positional
```
```suggestion
  of ParserKind.Command:
    ParameterKind.Command
  of ParserKind.Flag, FlagOptionalValue:
    ParameterKind.Flag
  of ParserKind.Positional..OptionalCatchAll:
    ParameterKind.Positional
```
Alternatively, you could also align the values horizontally.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+            proc(x: CommandId): bool = x == command
+          )
+    )
+    .get(otherwise = false)
```
```suggestion
  let parent = cli.parentOf(command)
  if parent.isSome:
    let default = cli.defaultCommandOf(parent.get)
    default.isSome and default.get == command
  else:
    false
```
It's shorter, easier to read (in my opinion), and a lot more efficient, as there's no copy of `cli`.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    .get(otherwise = false)
+
+func isValueOptional*(cli: Cli, flag: FlagId): bool =
+  ## Returns whether a value must be specified on the command line for `flag`.
```
```suggestion
  ## Returns whether a value for `flag` is optional on the command line.
```
The comment is inverted.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+func isOptional*(cli: Cli, positional: PositionalId): bool =
+  ## Returns whether a value must be specified on the command line for
+  ## `positional`.
```
```suggestion
  ## Returns whether a value for `positional` is optional on the command line.
```
The comment is inverted.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  cli.parser[ParameterId positional].kind in {CatchAll, OptionalCatchAll}
+
+proc helpFlagBuilder*[T](
+  cli: var Cli[T],
```
```suggestion
  cli: Cli[T],
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    .describe("display help message")
+
+proc addHelpFlag*[T](
+  cli: var Cli[T],
```
```suggestion
  cli: Cli[T],
```

alaviss replied:
I'm pretty sure this one has to be `var` :grin:

zerbina replied:
Oh, right, my bad.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+        .map(
+          proc (value: string): (CmdlineKind, string) =
+            (cmdValue, value)
+        )
```
```suggestion
        .map(proc (value: auto): auto = (cmdValue, value))
```
A soft suggestion.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  parse(cli, result, args)
+
+iterator flags*(cli: Cli, command: CommandId): FlagId =
+  ## Iterates through all added flags for `command`.
```
```suggestion
  ## Returns all flags for `command`.
```
A soft suggestion.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+      yield FlagId(i)
+
+iterator positionals*(cli: Cli, command: CommandId): PositionalId =
+  ## Iterates through all added positional parameters for `command`.
```
```suggestion
  ## Returns all positional parameters for `command`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+      yield PositionalId(i)
+
+iterator subcommands*(cli: Cli, command: CommandId): CommandId =
+  ## Iterates through all added subcommands for `command`.
```
```suggestion
  ## Returns all subcommands for `command`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+      yield CommandId(param)
+
+func flagsUsage*(cli: Cli, command: CommandId): string =
+  ## Produces an usage message for all flags registered for `command`.
```
```suggestion
  ## Produces a usage message for all flags registered for `command`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  ## For flags with required values, the placeholder is rendered as `VALUE`
+  ## if not set.
+  ##
+  ## A placeholder is not shown for flags with optional values if not set.
```
```suggestion
```
I'd leave it out, as `describe` already describes these cases.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    result.add "..."
+
+func positionalsUsage*(cli: Cli, command: CommandId): string =
+  ## Produces an usage message for all positionals registered for `command`.
```
```suggestion
  ## Produces a usage message for all positionals registered for `command`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+  command: CommandId,
+  rootName: string = cli.nameOf(RootCommand)
+): string =
+  ## Produces an usage message for the given command. The message is meant to
```
```suggestion
  ## Produces a usage message for the given command. The message is meant to
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+    result.add cli.displayOf(positional)
+
+func subcommandsUsage*(cli: Cli, command: CommandId): string =
+  ## Produces an usage message for all subcommands registered for `command`.
```
```suggestion
  ## Produces a usage message for all subcommands registered for `command`.
```

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2819 @@
+
+func prettifyError*[T](cli: Cli[T], error: ref ParseError): string =
+  ## Produce a pretty error message for `error`. The provided `error` must be
+  ## produced by `parse(cli)`.
```
```suggestion
  ## Produces a pretty error message for `error`. `error` must have been
  ## raised by `parse(cli)`.
```

zerbina reviewed (commented):
It's not a full review, but covers the most important parts, I think.

My main focus was on documentation and builder internals. I've only skimmed over the runner implementation, the tests, and the runnable examples.

I like the public API, though the longer I looked at it, the more I thought about whether an API that statically enforces constraints (e.g., via the builder creation procedure having required parameters for names, etc.) wouldn't be better (not implying that anything needs to change in this regard).

The code I've found fairly hard to read, largely due to the runnable example, but also because of the code style and the duplicated documentation. The workarounds for making tailcalls work also don't help.

There's also a bug that I've found:
```nim
var cli = commandBuilder(int).initCli()
cli.flagBuilder
  .name("b")
  .parser(
    proc(option: auto, value: auto, x: var int): auto =
      discard)
  .addTo(cli)

var i = 0
cli.run(i, @["-ab"])
# crashes due to an `unreachable` in `parse` being reached
```

My comments and suggestions are largely regarding comments, but I've also left some regarding naming and the implementation of builders.

alaviss commented:
> I like the public API, though the longer I looked at it, the more I thought about whether an API that statically enforces constraints (e.g., via the builder creation procedure having required parameters for names, etc.) wouldn't be better (not implying that anything needs to change in this regard).

Thanks.

The idea about builder creation statically enforcing the requirements did cross my mind, but there were a couple issues leading to them not being adopted:

- It prevents certain composing patterns: `helpFlagBuilder` couldn't just yield a builder with only the parser (it doesn't do this, but other people might).
- Some requirements can only be ergonomically set via an external procedure: a key example is a flag's parser, with its variants. It's certainly possible to have the parser be inline (i.e. `flagBuilder` vs `optionalFlagBuilder`), but it'd make overriding a parser looks weird (imagine `optionalFlagBuilder(...).parser()`).

There's a small runtime cost to offloading these checks there, but in general I think the API is a lot easier to use and compose like this.

alaviss committed:
cmdline: apply styling and docs suggestions

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

alaviss committed:
cmdline: cleanup and simplify positional and command builders

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2856 @@
+    ## Typeclass to support parsers returning either `Action` or nothing.
+
+  FlagParser*[T; R: MaybeAction] = proc (option, value: string, accumulator: var T): R
+    ## A parser for flag with name `option` and value `value`. The accumulator
```
I think `name` is a much better name for the parameter currently called `option`. 😆

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2856 @@
+proc `==`*(a, b: FlagId): bool {.borrow.}
+proc `==`*(a, b: PositionalId): bool {.borrow.}
+
+# FIXME: Put this in system.nim
+func drop[T](_: sink T) = discard
```
Any reason to not break this out into a PR now?

alaviss replied:
Without some magic in the compiler the `sink` might turn into a hidden copy if not careful.

I don't want to do the compiler part atm.

saem replied:
hmm, sounds like we'll need a `consume` like feature, rather than `sink`

saem commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2856 @@
+
+  cli.placeholder[ParameterId flag]
+
+func isDispatcher*(cli: Cli, command: CommandId): bool =
```
An alias of `hasSubcommands` seems appropriate, the dispatcher terminology is mentioned elsewhere, but I don't think it'd come to mind.

alaviss replied:
I think I might just phase the dispatcher terminology out.

zerbina commented:
> There's a small runtime cost to offloading these checks there, but in general I think the API is a lot easier to use and compose like this.

Yea, not being as composable as it is now is also something that came to my mind. My main issue with the dynamicness is less the runtime cost, but more that errors are only reported at run-time. There's also the increase in surface for implementation errors (due to the require error checking), but that's something the tests help with.

However, like I said, I think the builder-based API is fine as is.

alaviss committed:
cmdline: makes compatible with JS and VM backend

APIs that only work under native backends are placed under some
conditionals, allowing cmdline to compile with JS and VM backend.

However, due to bugs in the VM backend, cmdline will not function under
it.

alaviss committed:
cmdline: pluralification

alaviss committed:
cmdline: dispatcher -> hasSubcommand

alaviss committed:
cmdline: flag option -> name

zerbina commented on tests/stdlib/experimental/tcmdline.nim:
```diff
@@ -0,0 +1,2191 @@
+discard """
+description: "Tests for the cmdline module"
+targets: "c js vm"
+knownIssue.vm: "tables crashes with 'trying to access location outside of VM's memory'"
```
```suggestion
  description: "Tests for the cmdline module"
  targets: "c js vm"
  knownIssue.vm: '''
    `.tailcall` in combination with `sink` is not supported due to `vmgen` bug,
    causing an error at run-time 
  '''
```
This make the known-issue description more helpful to someone wanting to look into the issue.

zerbina commented on lib/experimental/cmdline.nim:
```diff
@@ -0,0 +1,2859 @@
+
+# FIXME: Change the detection to stdmsg once
+# https://github.com/nim-works/nimskull/pull/1547 is merged
+when declared(os.commandLineParams) and declared(stderr):
```
Why not move these procedures into a separate module?

alaviss replied:
The UX is horrible, since users would have to go import an another module to actually use cmdline (for typical usage).

If I use `export`, then the exported module is "invisible" in the generated docs, not to mention that it wouldn't simplify any of these conditionals.

zerbina replied:
> If I use `export`, then the exported module is "invisible" in the generated docs

What do you mean by invisible?

> not to mention that it wouldn't simplify any of these conditionals.

Why? There'd be no need for any conditional when `run` is in a separate module. If one imports the separate module with the VM/JS backend, compilation would simply fail.

alaviss replied:
> What do you mean be invisible?

The docs won't show up. It's a tradeoff I'd take for something like `cmdline/parsers`, since they're mostly helpers that you wouldn't normally interact with, not for something crucial to the typical usage flow.

> Why? There'd be no need for any conditional when `run` is in a separate module. If one imports the separate module with the VM/JS backend, compilation would simply fail.

I consider requiring to import two modules a deal breaker for cmdline, where the typical flow ends with user calling `run`.

zerbina replied:
> I consider requiring to import two modules a deal breaker for cmdline, where the typical flow ends with user calling `run`.

You wouldn't need to import two modules, just the one providing `run`, as that also exports `cmdline`. The routines not being listed directly on the separate module's documentation page can be easily helped with by linking to `cmdline` at the top, in my opinion.

I'd prefer there being no `when` workarounds here, given the current state of multi-platform/backend support, but in the end, I don't consider the presence of `when` a blocker for this PR.

alaviss committed:
tcmdline: update knownIssue metadata

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

alaviss commented:
Rebased to import #1547

zerbina reviewed (approved):
From my side, the changes look good. Nice work!

The test coverage (from both the tests and runnable examples) looks great, though I've had trouble following the code due to repetitive shape and sheer volume, thus making it likely that I've missed something in this area.

In the future, please use merge instead of rebase, at least when there are review comments already -- the rebase makes it hard to keep track of what was reviewed already.

I'm not merging the PR yet, so that @saem can take another look.

saem commented on tests/stdlib/experimental/tcmdline.nim:
```diff
@@ -0,0 +1,2194 @@
+discard """
+  description: "Tests for the cmdline module"
+  targets: "c js vm"
```
noice!

saem commented:
merge

nice job, @alaviss!

saem commented:
/merge

nice job, @alaviss!
